src/handlers.rs

```diff
@@ -1,1139 +1,1406 @@
-    // Return
-    #[derive(Debug, Deserialize, Serialize)]
-    struct Response {
-        status_code: u16,
-        message: models::Message,
+    if let Err(e) = tx.commit() {
+        error!("Failed to commit message: {}", e);
+        return Err(warp::reject::custom(Error::DatabaseFailedInternally));
     }
-    let response = Response { status_code: StatusCode::OK.as_u16(), message };
-    return Ok(warp::reply::json(&response).into_response());
-}
 
-fn get_last_5_messages(
-    public_key: &str, pool: &storage::DatabaseConnectionPool,
-) -> Result<Vec<models::Message>, Rejection> {
-    let conn = pool.get().map_err(|_| Error::DatabaseFailedInternally)?;
-    let raw_query = format!(
-        "SELECT id, public_key, timestamp, data, signature FROM {} WHERE public_key = (?1) ORDER BY timestamp DESC LIMIT 5",
-        storage::MESSAGES_TABLE
-    );
-    let mut query = conn.prepare(&raw_query).map_err(|_| Error::DatabaseFailedInternally)?;
-    let rows = match query.query_map(params![public_key], |row| {
-        Ok(models::Message {
-            server_id: row.get(0)?,
-            public_key: row.get(1)?,
-            timestamp: row.get(2)?,
-            data: row.get(3)?,
-            signature: row.get(4)?,
-        })
-    }) {
-        Ok(rows) => rows,
-        Err(e) => {
-            error!("Couldn't get last 5 messages due to error: {}.", e);
-            return Err(warp::reject::custom(Error::DatabaseFailedInternally));
-        }
-    };
-    return Ok(rows.filter_map(|result| result.ok()).collect());
+    let response = json!({ "status_code": StatusCode::OK.as_u16(), "message": message });
+    Ok(warp::reply::json(&response).into_response())
 }
 
-/// Returns either the last `limit` messages or all messages since `from_server_id, limited to `limit`.
-pub fn get_messages(
-    query_params: HashMap<String, String>, auth_token: &str, pool: &storage::DatabaseConnectionPool,
-) -> Result<Vec<models::Message>, Rejection> {
-    // Check authorization level
-    let (has_authorization_level, _) =
-        has_authorization_level(auth_token, AuthorizationLevel::Basic, pool)?;
-    if !has_authorization_level {
-        return Err(warp::reject::custom(Error::Unauthorized));
-    }
-    // Get a database connection
-    let conn = pool.get().map_err(|_| Error::DatabaseFailedInternally)?;
-    // Unwrap query parameters
-    let from_server_id: i64;
+// TODO FIXME: The paging mechanism here is really odd: you can either get the last 256 messages,
+// *or* you can start getting messages from the beginning of time and walk forward.  This is
+// really, really inefficient, though, because if you join a very old room you could end up having
+// to load the entire history from the beginning of time if you want to scroll back before the
+// 256th-last message.
+//
+// (The same applies to get_deleted_messages, below).
+
+fn get_messages_params(query_params: &HashMap<String, String>) -> (Option<i64>, u16) {
```

Comment:
Telegram looks like they have 3 methods of retrieving messages:
- Limit provided as an arg from the client
- Offset skips a number of elements since beginning of list
- Offset from ID calls like loading 20 messages, older than message with ID MSGID: `messages.getHistory({offset_id: MSGID, add_offset: 0, limit: 20})`

https://core.telegram.org/api/offsets